### ISSUE TITLE
nvmf dif_insert_or_strip flag is transport option

### ISSUE THREAD

**Issue by changpe1** [spdk/spdk#2438]
# Bug report

Currently we set `dif_insert_or_strip` when creating transports, which means the flag is based on transport.

## Expected Behavior

The flag should be enabled via IO flags in NVMe command when PRACT=1 and metadata size is 8.

## Possible Solution

Implement spdk_nvmf_request_get_dif_ctx() based on NVMe command IO flags and bdev capabilities.

**Comment by changpe1** [comment]
Currently, we only report extended LBA format to initiator when `dif_insert_or_strip` is false, when `dif_insert_or_strip` is enabled, we will only report sector size to initiator, the initiator has no way to know the behaviors in target.  I know nvmf doesn't support FORMAT command now, let's have a discussion here.

**Comment by mhae** [comment]
It would be good if we could put some smarts in the target to figure out the best approach on how to handle the IO.

If the initiator sends PRACT=1 with 512b SGEs, then we could **either** pass this on to the backend (e.g. bdev_nvme.c) or do the insert/strip in NVMF.
It really depends on the capabilities of the actual backend. 

Now, there are probably quite a number of ripple effects to support this... For example, if the NVMe device should do the insert/strip then it is unclear to me if the NVMF/bdev/bdev_nvme layer would work correctly with 520b sectors, or if these layers would require 512b sectors.

**Comment by changpe1** [comment]
> It would be good if we could put some smarts in the target to figure out the best approach on how to handle the IO.
> 
> If the initiator sends PRACT=1 with 512b SGEs, then we could **either** pass this on to the backend (e.g. bdev_nvme.c) or do the insert/strip in NVMF. It really depends on the capabilities of the actual backend.
> 
> Now, there are probably quite a number of ripple effects to support this... For example, if the NVMe device should do the insert/strip then it is unclear to me if the NVMF/bdev/bdev_nvme layer would work correctly with 520b sectors, or if these layers would require 512b sectors.

Currently initiator NVMe bdev and the backend NVMe bdev isn't connected directly, nvmf library does the connection, which means PRACT and PRCHK flags are all ignored.

Backend NVMe bdev is formatted to 520B.
case 1: if `dif_insert_or_strip` is false, nvmf library will report 520B extended LBA format to initiator but with DIF flag set to UNSUPPORTED.
case 2: if `dif_insert_or_strip` is true, nvmf library will report 512B to initiator, then nvmf library will do data buffer insert and strip.

**Comment by shuhei-matsumoto** [comment]
@changpe1 

> Backend NVMe bdev is formatted to 520B.
> case 1: if dif_insert_or_strip is false, nvmf library will report 520B extended LBA format to initiator but with DIF flag set to UNSUPPORTED.
> case 2: if dif_insert_or_strip is true, nvmf library will report 512B to initiator, then nvmf library will do data buffer insert and strip.

For the case 1, Before the patch https://review.spdk.io/gerrit/c/spdk/spdk/+/12165 , spdk_bdev_get_dif_type() was exposed to the initiator. How we can restore this feature?

Do you want lib/nvmf to support PRACT per I/O? If PRACT per I/O is supported and transport->dif_insert_or_strip is disabled, can lib/nvmf expose spdk_bdev_get_dif_type() to the initiator?

Do you really require lib/nvmf to support FORMAT command?

Please check the following comments too.

@mhae

> It would be good if we could put some smarts in the target to figure out the best approach on how to handle the IO.
> 
> If the initiator sends PRACT=1 with 512b SGEs, then we could either pass this on to the backend (e.g. bdev_nvme.c) or do the insert/strip in NVMF.
> It really depends on the capabilities of the actual backend.
> 
> Now, there are probably quite a number of ripple effects to support this... For example, if the NVMe device should do the insert/strip then it is unclear to me if the NVMF/bdev/bdev_nvme layer would work correctly with 520b sectors, or if these layers would require 512b sectors.

The upper layer can pass spdk_bdev_ext_io_opts to API spdk_bdev_readv/writev_blocks_ext().
Can we add pract bit into spdk_bdev_ext_io_opts?
Then each bdev module can decide if it insert/strip DIF itself or pass pract down to the next bdev module.
For the NVMe bdev module, it can choose either insert/strip DIF itself or let NVMe SSD do insert/strip DIF by passing pract.

And lib/nvmf has the following options:
- dif_insert_or_strip is enabled: DIF is not exposed to the initiator. lib/nvmf insert/strip DIF itself.
- dif_insert_or_strip is disabled, per_io_dif_insert_or_strip is disabled, I/O with PRACT is sent: bring PRACT into the bdev.
- dif_insert_or_strip is disabled, per_io_dif_insert_or_strip is enabled, I/O with PRACT is sent: lib/nvmf insert/strip DIF itself.

How do you think, any feedback?

-Shuhei

**Comment by changpe1** [comment]
> 

I prefer to deprecate `dif_insert_or_strip` flag in nvmf library, and follow the NVMe specification to check PRACT and other flags, but it's up to us to determine full support of the specification or part support of the specification.

**Comment by shuhei-matsumoto** [comment]
@changpe1 By my old memory, the approach something like dif_insert_or_strip is very popoular.

**Comment by shuhei-matsumoto** [comment]
So please do not deprecate dif_insert_or_strip flag and keep it forward.

**Comment by shuhei-matsumoto** [comment]
Can we revert https://review.spdk.io/gerrit/c/spdk/spdk/+/12165  if lib/nvmf support PRACT?

**Comment by changpe1** [comment]
> Can we revert https://review.spdk.io/gerrit/c/spdk/spdk/+/12165 if lib/nvmf support PRACT? Please continue discussion on slack or here.

No, this is the intention to work with `dif_insert_or_strip` because we don't report this capability to initiator, so that PRACT is ignored in nvmf library.

**Comment by shuhei-matsumoto** [comment]
@changpe1 Thank you. What I want to ask was not clear. lib/nvmf want to support DIF passthrough in future. (In this case,  NVMe-oF transport just verifies DIF.) Hence, I want to know your idea about PRACT. How PRACT support in lib/nvmf look like? If you have concrete idea, will you share it?

**Comment by changpe1** [comment]
> @changpe1 Thank you. What I want to ask was not clear. lib/nvmf want to support DIF passthrough in future. (In this case, NVMe-oF transport just verifies DIF.) Hence, I want to know your idea about PRACT. How PRACT support in lib/nvmf look like? If you have concrete idea, will you share it?

Existing code is OK, at least it doesn't violate the specification.
I don't have a good idea to keep `dif_insert_or_strip` and meanwhile still to support PRACT and PRCHK flags.
We need a decision to whether to support initiator PRACT and PRCHK flags or not?

**Comment by shuhei-matsumoto** [comment]
@changpe1 Even now, DIF context is per I/O. We initialize DIF context based on the information from the bdev. However, we can overwrite PRACT and PRCHK of the initialized DIF context. And we can change current dif_insert_or_strip flag to enum, per_bdev_enable, per_io_enable_if_pract, disable. Let's discuss more after you are ready.

**Comment by changpe1** [comment]
> @changpe1 Even now, DIF context is per I/O. We initialize DIF context based on the information from the bdev. However, we can overwrite PRACT and PRCHK of the initialized DIF context. And we can change current dif_insert_or_strip flag to enum, per_bdev_enable, per_io_enable_if_pract, disable. Let's discuss more after you are ready.

Yeah, but  this is done when `dif_insert_or_strip` is TURE in target side, but we don't report this capability to initiator, so initiator should not set PRCHK and PRACT for this case.

For supporting the feature you mentioned above, we should always report this capability to initiator, that's my above patch did at the beginning.

**Comment by mhae** [comment]
@shuhei-matsumoto, yes, your summary is correct.

**Comment by benlwalker** [comment]
So I'm trying to put together a more holistic view of all of the extended block size support. This is going to be long, but maybe we can convert it into some documentation and hopefully it helps to clarify the situation and what our next steps should be.

First, there's a lot of different cases here that we need to think through:
* The front-end namespace exported by NVMe-oF can either be formatted with protection information or not.
* Individual NVMe-oF commands can set the PRACT and PRCHK values to modify all sorts of behaviors.
* The back-end bdev can either be formatted with protection information or not.
* Multiple namespaces can be accessed by a single queue pair, and those namespaces do not need to share a format.

Let's start by thinking only about the front-end queues and exported namespaces within a subsystem. Each queue allocates memory to hold in-capsule data and the transport allocates data buffers for general use per group. Since a queue communicates with a subsystem, that means different namespaces can have different formats, but share the same queue. This is problematic for these memory pools - especially the in-capsule data buffers. The allocations need to handle the worst case for any namespace exported by the subsystem. In fact, this seems to be all that `dif_insert_or_strip` actually controls.

Namespaces can be added and removed from a subsystem dynamically, at any time. So these queues can't simply query all of the namespaces and make a decision about how much in-capsule data space to allocate. Whether the queue should plan to support extended block sizes needs to be provided as a configuration parameter instead. We currently use the transport-wide option `dif_insert_or_strip`. That's probably fine, but it might be better done per-subsystem. It could also benefit from a different name such as `extended_block_support` which has a reversed value from `dif_insert_or_strip`.

Now we can look at the interaction between whether the front-end is capable of supporting extended block sizes and what the bdev backing the namespace can actually do. There's a 2x2 matrix of options here broadly speaking (although lots of complexity on the specific formats of the PI data).

1. (`extended_block_support` = true, bdev supports metadata) => exported namespace has extended blocks
2. (`extended_block_support` = true, bdev does not support metadata) => exported namespace does not have extended blocks
3. (`extended_block_support` = false, bdev supports metadata) => exported namespace does not have extended blocks
4. (`extended_block_support` = false, bdev does not support metadata) => exported namespace does not have extended blocks

Cases 2 and 4 are easy. Case 3 needs more thought. I haven't investigated whether it's always valid to send commands with PRACT and PRCHK set to do nothing with metadata. If it is, that's what we should do. Just ignore that the metadata exists.

Case 1 is then where we'll focus because that's the hard one. First, I think it's clear that the exported namespace's metadata size should always match the backing bdev to eliminate complexity. Then it's just a matter of passing metadata through, if it's provided in the command.

If the command contains a PRACT that says to do automatic insert/strip of metadata, then we have to figure out what to do. I think the options are to either add flags to the `struct spdk_bdev_ext_io_opts` where we can pass down a setting that says to do automatic PI insert/strip inside of the bdev itself. This would be the most efficient implementation. Or we could do this entirely via software emulation in `lib/nvmf`, but I am leaning toward the flags in `struct spdk_bdev_ext_io_opts`.

Before I go any further, does this all make sense so far?

**Comment by changpe1** [comment]
@benlwalker good summary.
For Case 3, I prefer to expose the extended blocks to initiator, if so, we don't need the control flag `extended_block_support`, when doing IOs, we can follow the NVMe spec to determine the data block contains extended data buffer or not, actually we already do it in the nvme driver.

And for supporting PRCHK flags, I think existing emulation in `lib/nvmf` is fine, we just need to respect the PRCHK flags in each command.

Most important of all, should we move forward for this issue? Existing implementation is OK to me.

**Comment by shuhei-matsumoto** [comment]
We have to keep the current `dif_insert_or_strip`. This is necessary to offload DIF insert/strip into NIC hardware or does it by software.
Thank you so much @benlwalker too.

Ben says `extended_block_support` has a reversed value from `dif_insert_or_strip`.
I think we still need either one of `extended_block_support` or `dif_insert_or_strip`.
We cannot merge the case 3 into the case 2.

I believe
1a. `extended_block_support` = true, bdev supports metadata, IO has PRACT
is different from
3. `extended_block_support` = false, bdev supports metadata

And I agree with Ben that for 1a we can use the flags in `struct spdk_bdev_ext_io_opts`.

And For 1, we may want NIC hardware to support DIF passthrough, that is, only verify DIF at the entrance/exit.

@changpe1 for us, it has a value to move forward this issue, but I am not certain if I can work on this issue until completion yet.

**Comment by mhae** [comment]
@changpe1 , regarding usefulness: This feature would have saved us a ton of hassle moving between 512 and 520 disks.

**Comment by benlwalker** [comment]
After discussion, I think SPDK is currently only missing 1 feature (that could be used in two places) - the PRACT/PRCHK equivalents on `struct spdk_bdev_ext_io_opts`. I think we do need a way to query whether the bdev supports generating/checking it's own PI (only NVMe devices will). Then, in case 1 if the feature is available, we can translate the NVMe PRACT/PRCHK flags into these bdev layer options and pass the I/O down. Similarly, for case 3 we should also just use these features, rather than generating our own PI information.

We do still need to be able to generate our own PI information for case 3 when the bdev is not an NVMe device. This is the case where you'd want to use the RDMA NIC offloads to generate it as the data is received.

@mhae - does that make sense? I wasn't sure which feature you meant would have saved you work.

**Comment by mhae** [comment]
@benlwalker , sorry, I wasn't clear... I think what you described is what would have helped... Disk is formatted 520 and initiator sends PRACT=1 with 512 sectors. The disk would do the insert/strip in this case. 
Likewise, if initiator sends PRACT=0 with 520 sectors, the disk would NOT do the insert/strip (initiator calculated the checksum).

**Comment by changpe1** [comment]
> @benlwalker , sorry, I wasn't clear... I think what you described is what would have helped... Disk is formatted 520 and initiator sends PRACT=1 with 512 sectors. The disk would do the insert/strip in this case. Likewise, if initiator sends PRACT=0 with 520 sectors, the disk would NOT do the insert/strip (initiator calculated the checksum).

This is exactly what physical NVMe SSD behave in the initiator side. Metadata size 8 is a special case, when PRACT=1, initiator should use 512B, and use 520 when PRACT=0, and for metadata size greater than 8, extended block size is used for all cases.

**Comment by shuhei-matsumoto** [comment]
Hi,

A patch series to expose end-to-end data protection to host is on gerrit now.
It starts from https://review.spdk.io/gerrit/c/spdk/spdk/+/22918

The idea is to utilize existing DIF insert/strip feature of NVMe-oF transports to support NVMe PRACT.
Enabling DIF insert/strip is required to expose DIF to host for now.
However, I am adding NVMe PRACT passthrough now.
Hopefully the limitation will be removed soon.

And more missing features will be added soon.

**Comment by mhae** [comment]
Hey Shuhei, let me know when the series is ready for review and I'll take a
look at it.

Michael
